# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [1]:
import numpy as np
import pandas as pd

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents

In [2]:
salaries = pd.read_csv('../Current_Employee_Names__Salaries__and_Position_Titles.csv')

Examine the `salaries` dataset using the `head` function below.

In [3]:
salaries.head(2)

Name                              Job Titles Department  \
0  AARON,  JEFFERY M                                SERGEANT     POLICE   
1    AARON,  KARINA   POLICE OFFICER (ASSIGNED AS DETECTIVE)     POLICE   

  Full or Part-Time Salary or Hourly  Typical Hours  Annual Salary  \
0                 F           Salary            NaN       101442.0   
1                 F           Salary            NaN        94122.0   

   Hourly Rate  
0          NaN  
1          NaN

In [4]:
salaries.shape

(33183, 8)

In [5]:
salaries.isna().sum()

Name                     0
Job Titles               0
Department               0
Full or Part-Time        0
Salary or Hourly         0
Typical Hours        25161
Annual Salary         8022
Hourly Rate          25161
dtype: int64

We see from looking at the `head` function that there is quite a bit of missing data. Let's examine how much missing data is in each column. Produce this output in the cell below

In [6]:
salaries.isnull().values.sum()

58344

Let's also look at the count of hourly vs. salaried employees. Write the code in the cell below

In [7]:
salaries['Salary or Hourly'].unique()

array(['Salary', 'Hourly'], dtype=object)

In [8]:
salaries['Salary or Hourly'].value_counts()

Salary    25161
Hourly     8022
Name: Salary or Hourly, dtype: int64

What this information indicates is that the table contains information about two types of employees - salaried and hourly. Some columns apply only to one type of employee while other columns only apply to another kind. This is why there are so many missing values. Therefore, we will not do anything to handle the missing values.

There are different departments in the city. List all departments and the count of employees in each department.

In [9]:
salaries.Department.value_counts()

POLICE                   13414
FIRE                      4641
STREETS & SAN             2198
OEMC                      2102
WATER MGMNT               1879
AVIATION                  1629
TRANSPORTN                1140
PUBLIC LIBRARY            1015
GENERAL SERVICES           980
FAMILY & SUPPORT           615
FINANCE                    560
HEALTH                     488
CITY COUNCIL               411
LAW                        407
BUILDINGS                  269
COMMUNITY DEVELOPMENT      207
BUSINESS AFFAIRS           171
COPA                       116
BOARD OF ELECTION          107
DoIT                        99
PROCUREMENT                 92
INSPECTOR GEN               87
MAYOR'S OFFICE              85
CITY CLERK                  84
ANIMAL CONTRL               81
HUMAN RESOURCES             79
CULTURAL AFFAIRS            65
BUDGET & MGMT               46
ADMIN HEARNG                39
DISABILITIES                28
TREASURER                   22
HUMAN RELATIONS             16
BOARD OF

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [10]:
sample=salaries['Hourly Rate']
sample.dropna(inplace=True)

In [11]:
sample

4        19.86
6        46.10
7        35.60
10        2.65
18       17.68
         ...  
33164    46.10
33168    17.68
33169    35.60
33174    46.35
33175    48.85
Name: Hourly Rate, Length: 8022, dtype: float64

In [12]:
salaries.describe()

Typical Hours  Annual Salary  Hourly Rate
count    8022.000000   25161.000000  8022.000000
mean       34.507604   86786.999790    32.788558
std         9.252077   21041.354602    12.112573
min        10.000000    7200.000000     2.650000
25%        20.000000   76266.000000    21.200000
50%        40.000000   90024.000000    35.600000
75%        40.000000   96060.000000    40.200000
max        40.000000  300000.000000   109.000000

In [13]:
# H0 is significantly different from $30/hr means > 30 or <30 USD/HR
# H1 : is close to $30/hr
import numpy as np
from scipy.stats import ttest_1samp

ttest_1samp(sample, 30)

Ttest_1sampResult(statistic=20.6198057854942, pvalue=4.3230240486229894e-92)

Answer: pvalue is close to 0, it is not in the 95% confidence interval, so we reject the Null Hypothesis.

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are higher than last year's mean of $86000/year a year for all salaried employees. Test this one sided hypothesis using a 95% confidence interval.

Hint: A one tailed test has a p-value that is half of the two tailed p-value. If our hypothesis is greater than, then to reject, the test statistic must also be positive.

In [14]:
salaries['Annual Salary'].groupby(salaries.Department).mean()

Department
ADMIN HEARNG             78683.692308
ANIMAL CONTRL            66197.612903
AVIATION                 78750.549324
BOARD OF ELECTION        53548.149533
BOARD OF ETHICS          95061.000000
BUDGET & MGMT            91836.545455
BUILDINGS                99246.761041
BUSINESS AFFAIRS         80577.073171
CITY CLERK               70899.392857
CITY COUNCIL             64482.566801
COMMUNITY DEVELOPMENT    88430.778325
COPA                     76150.034483
CULTURAL AFFAIRS         89566.551724
DISABILITIES             83879.571429
DoIT                     99978.060606
FAMILY & SUPPORT         79895.762195
FINANCE                  73781.255814
FIRE                     97958.600811
GENERAL SERVICES         83545.057116
HEALTH                   85557.501031
HUMAN RELATIONS          94819.500000
HUMAN RESOURCES          80230.560000
INSPECTOR GEN            79874.758621
LAW                      85958.321526
LICENSE APPL COMM        84420.000000
MAYOR'S OFFICE           98186.142857
O

In [15]:
#salaries['Annual Salary'].groupby(salaries.Department).mean().nlargest(5)
subsetDataFrame = salaries[salaries['Department'] == 'POLICE'] #.sample(n=700)
subsetDataFrame['Annual Salary']

0        101442.0
1         94122.0
9         93354.0
11        84054.0
12        87006.0
           ...   
33177     72510.0
33178     72510.0
33179     48078.0
33180     90024.0
33181     93354.0
Name: Annual Salary, Length: 13414, dtype: float64

In [16]:
ttest_1samp(subsetDataFrame['Annual Salary'].dropna(), 86000)

Ttest_1sampResult(statistic=3.081997005712994, pvalue=0.0020603403550965137)

Ttest_1sampResult(statistic=-0.8523576761115091, pvalue=0.3960729175654153 for the sample of n=100
Ttest_1sampResult(statistic=1.4379954758257858, pvalue=0.15358893333559012 for the sample of n=100
Ttest_1sampResult(statistic=0.5324702769298395, pvalue=0.5949943555859245 for the sample of n=200
Ttest_1sampResult(statistic=0.2137835585476616, pvalue=0.8309347247341368 for the sample of n=200
Ttest_1sampResult(statistic=0.2751969310542946, pvalue=0.7832788559861887 for the sample of n=500
Ttest_1sampResult(statistic=0.6705540571238218, pvalue=0.5027261438181817 for the sample of n=700
Ttest_1sampResult(statistic=1.2229781912974422, pvalue=0.2216584242991598 for the sample of n=900
Ttest_1sampResult(statistic=2.168288313541497, pvalue=0.0303722185414443 for the sample of n=1000
2000 sample sorts out the NAN values

To plot the data to understand the pvalue

2) get the random values out of the series of subsetDataFrame['Annual Salary']

After having dropped the NAN value of the Hourly employees in the Police department the result is
Ttest_1sampResult(statistic=3.081997005712994, pvalue=0.0020603403550965137)
pvalue is less than 0,05 - it is positif value => positif tail ?
H0 of the mean salary is rejected, as the pvalue is not in the confidence interval

Using the `crosstab` function, find the department that has the most hourly workers. 

In [17]:
subset_hourly= salaries[salaries['Salary or Hourly'] == 'Hourly']

In [18]:
x=pd.crosstab(subset_hourly['Salary or Hourly'], subset_hourly.Department, dropna=False).max().nlargest(3)
x

Department
STREETS & SAN    1862
WATER MGMNT      1513
OEMC             1273
dtype: int64

In [19]:
biggest=salaries[salaries['Department'] == 'STREETS & SAN']

The workers from the department with the most hourly workers have complained that their hourly wage is less than $35/hour. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [20]:
# H0 is the Hyp that the the hourly wage is less then 35 USD/Hour.
ttest_1samp(biggest['Hourly Rate'].dropna(), 35)

Ttest_1sampResult(statistic=-9.567447887848152, pvalue=3.3378530564707717e-21)

In [ ]:
Ttest_1sampResult(statistic=-9.567447887848152, pvalue=3.3378530564707717e-21)
pvalue 0 % is in the range of out of the condifence level, well the workers are not paid less then 35 USD/H

In [25]:
biggest.describe()

Typical Hours  Annual Salary  Hourly Rate
count    1862.000000     336.000000  1862.000000
mean       37.137487   83019.250000    33.728378
std         8.777897   21254.991182     5.735242
min        10.000000   31872.000000    19.500000
25%        40.000000   70056.000000    35.600000
50%        40.000000   84168.000000    35.600000
75%        40.000000   97740.000000    36.210000
max        40.000000  154428.000000    53.100000

# Challenge 3 - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [22]:
# Your code here:


Now construct the 95% confidence interval for all salaried employeed in the police in the cell below.

In [23]:
# Your code here:


# Challenge 4 - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [24]:
# Your code here:
